docs/dag-sharding-rfc.md

```diff
@@ -0,0 +1,182 @@
+
+The format we use to encode file metadata in HTTP multipart requests can be mimiced from ipfs and should be standardized as much as possible for intuitive use of the API endpoint.  
+
+Designing a streaming blockstore for outputting block nodes begs similar questions posed in the user API section.  Multiple methods exist for streaming blocks out and we will need to investigate our desired assumptions and performance goals further to make a decision here.  Note that in addition to the data transfer suggestions visited in the earlier section, here we have another option: pass data to the sharding cluster component by instantiating it as a component of the output blockstore used in the importer.
+
+
+### Sharding cluster component
+
+At a high level the sharding component takes in a stream of dag-node blocks, maintains the necessary data to build a shard node of the Cluster DAG, and forwards the stream of blocks to a peer for pinning.  When blocks arrive the sharding process records the cids which are needed for constructing links between the shard node and the blocks of the DAG being sharded.  This is all the sharding component must do before forwarding blocks to other peers.  The peer receiving the shard is calculated once by querying the allocator.  The sharding component must be aware of how much space peers have for storing shards and will maintain state for determining when a shard is complete and the receiving peer needs to be recalculated.  The actual forwarding can be accomplished by periodically calling an RPC that runs DAG API `put` commands with processed blocks on the target peer.  After all the blocks of a shard are transmitted the relevant subgraph of the Cluster DAG is constructed and sent as a final transmission to the allocated peer.  Upon receiving this final transmission the allocated peer recursively pins the node of the Cluster DAG that it receives, ensuring that all blocks of the shard stay in its ipfs repo.  When all shards are constructed and allocated the final root node pointing to all shard node is pinned (non-recursively) throughout the cluster.
```

Comment:
We do need to make sure that ipfs GC does not run between the moment when we add the first block of a shard and the moment we pin the shard. This is workaroundable by pinning every block individually (and later unpinning leaving only the recursive), but probably at a high price.